File: 백준/Silver/4396. 지뢰 찾기/지뢰 찾기.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const n = Number(input[0])

const mines = input.slice(1, 1 + n).map((row) => row.split(""))
const choices = input.slice(1 + n).map((row) => row.split(""))
const graph = mines.map((row) => row.map(() => "."))

// 지뢰 눌렀는지 확인
const pressMines = (mines, choices) => {
  for (let i = 0; i < n; i += 1) {
    for (let j = 0; j < n; j += 1) {
      if (mines[i][j] === "*" && choices[i][j] === "x") {
        return true
      }
    }
  }
  return false
}
const start = (graph, mines, choices) => {
  let flag = pressMines(mines, choices)
  getGraph(graph, mines, choices)
  if (flag) {
    getMines(graph, mines)
  }
  console.log(graph.map((row) => row.join("")).join("\n"))
}

// 1. 지뢰 누른 경우
const getMines = (graph, mines) => {
  for (let i = 0; i < n; i += 1) {
    for (let j = 0; j < n; j += 1) {
      if (mines[i][j] === "*") {
        graph[i][j] = "*"
      }
    }
  }
}
// 2. 지뢰 안누른 경우

const direction = [
  [-1, 0],
  [-1, 1],
  [0, 1],
  [1, 1],
  [1, 0],
  [1, -1],
  [0, -1],
  [-1, -1],
]
const check = (mines, x, y) => {
  let cnt = 0
  direction.forEach(([nx, ny]) => {
    if (x + nx < 0 || x + nx >= n || y + ny < 0 || y + ny >= n) {
      return
    }
    if (mines[x + nx][y + ny] === "*") {
      cnt += 1
    }
  })
  return cnt
}
const getGraph = (graph, mines, choices) => {
  for (let i = 0; i < n; i += 1) {
    for (let j = 0; j < n; j += 1) {
      if (choices[i][j] === "x") {
        graph[i][j] = check(mines, i, j)
      }
    }
  }
}
start(graph, mines, choices)

File: 백준/Silver/4396. 지뢰 찾기/지뢰 찾기.py
n = int(input())
mine = [list(input()) for _ in range(n)]
board = [list(input()) for _ in range(n)]
result = [['.'] * n for _ in range(n)]
dx = [-1, -1, -1, 0, 0, 1, 1, 1]
dy = [-1, 0, 1, -1, 1, -1, 0, 1]
 
for x in range(n):
    for y in range(n):
        if mine[x][y] == '.' and board[x][y] == 'x':    # 지뢰가 없으면서 열린 칸
            cnt = 0
            for i in range(8):
                nx = x + dx[i]
                ny = y + dy[i]
 
                if nx < 0 or ny < 0 or nx >= n or ny >= n:
                    continue
 
                if mine[nx][ny] == '*':
                    cnt += 1
            result[x][y] = cnt    # 19번의 조건 절을 통과하지 않아도 cnt 적용
 
        if mine[x][y] == '*' and board[x][y] == 'x':    # 지뢰가 있는 칸이 열렸다면
            for a in range(n):
                for b in range(n):
                    if mine[a][b] == '*':    # 첫번째 입력 중 지뢰가 있는 칸이면
                        result[a][b] = '*'    # 지뢰가 있는 모든 칸을 별표로 표시
                        
for i in range(n):
    for j in range(n):
        print(result[i][j], end='')
    print()